## Titanic Dataset - Data Cleaning and Exploratory Data Analysis (EDA)

In this notebook, we focus on **data preparation**, **cleaning**, and **exploratory data analysis** for the **Titanic dataset**, a well-known dataset used to predict passenger survival based on demographic and travel-related attributes.

Good data preprocessing is essential for accurate and meaningful analysis. Here, we handle common data issues such as **missing values, duplicates, and inconsistent categorical variables**, while also performing **feature engineering** to create new variables that can provide deeper insights during analysis.


We start by importing essential Python libraries for data handling, analysis, and visualization:

- `pandas` for structured data operations.

- `numpy` for numerical operations.

- `os` for interacting with the operating system and directory structures.

In [3]:
# Import libraries

import pandas as pd
import numpy as np
import os

## 1. Define and Create Directory Paths

To ensure reproducibility and organized storage, we programmatically create directories if they don't already exist for:

- **raw data**
- **processed data**
- **results**
- **documentation**

These directories will store intermediate and final outputs for reproducibility.


In [4]:
# Get working directory
current_dir = os.getcwd()
# Go one directory up to the root directory
project_root_dir = os.path.dirname(current_dir)
# define paths to the data files
data_dir = os.path.join(project_root_dir, 'data')
raw_dir = os.path.join(data_dir, 'raw')
processed_dir = os.path.join(data_dir, 'processed')
# Define paths to the results folder
results_dir  = os.path.join(project_root_dir, 'results')
# Define paths to the docs folder
docs_dir = os.path.join(project_root_dir, 'docs')

# create directories if they do not exist 
os.makedirs(raw_dir, exist_ok = True )
os.makedirs(processed_dir, exist_ok = True )
os.makedirs(results_dir, exist_ok = True)
os.makedirs(docs_dir, exist_ok = True)

## 2. Read in the Data

We load the **Titanic Dataset** as a CSV file (rainic.csv`). The dataset contains passenger information such as demographics, ticket details, and survival status. It is read into a Pandas DataFrame for analysis.

# 🔑 Key Considerations
- **File format:** The dataset is stored as a CSV file, which can be easily loaded into Python using Pandas.  
- **Handling missing data:** Some columns, like `Age` and `Cabin`, contain missing values that should be addressed during cleaning.  
- **Data types:** Ensure that numeric columns (e.g., `Age`, `Fare`) are correctly recognized as numeric types, and categorical columns (e.g., `Sex`, `Embarked`) as categorical types for analysis.  
- **Previewing the data:** Displaying the first 10 rows using `head(10)` helps confirm that the data has been loaded correctly and gives an initial look at the dataset’s structure.  
- **Exploring columns:** Understanding each column’s meaning (e.g., `SibSp`, `Parch`, `Pclass`) is important for feature engineering and analysis.



After loading, we inspect the first few rows to understand the structure and content of the dataset.


In [5]:
train_df = pd.read_csv(r"C:\\Users\\Tasha\\Desktop\\TITANIC_ML\\data\\raw\\train.csv")
train_df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

We also inspect the dataset's shape. We see that the data has *891* rows and *12* columns.


In [6]:
train_df.shape

(891, 12)

In addition, we check the data types using .info.

In [7]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


## 3. Data Cleaning
### 3.1. Assign Proper Column Names

The Titanic dataset already includes column headers, but it’s important to **verify their correctness** and ensure they are **descriptive and consistent**.  
Proper column names improve **readability** and **interpretability** during analysis.

For reference, the main columns are:  
- `PassengerId` – Unique identifier for each passenger  
- `Survived` – Survival status (0 = No, 1 = Yes)  
- `Pclass` – Passenger class (1 = 1st, 2 = 2nd, 3 = 3rd)  
- `Name` – Passenger name  
- `Sex` – Gender  
- `Age` – Age in years  
- `SibSp` – Number of siblings/spouses aboard  
- `Parch` – Number of parents/children aboard  
- `Ticket` – Ticket number  
- `Fare` – Passenger fare  
- `Cabin` – Cabin number  
- `Embarked` – Port of embarkation (C = Cherbourg, Q = Queenstown, S = Southampton)


In [8]:
train_df.columns = ["PassengerId", "Survived", "Pclass", "Name", "Sex", "Age", "SibSp", "Parch", "Ticket", "Fare", "Cabin", "Embarked"]

We inspect again to see whether they are properly assigned

In [9]:
train_df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

### 3.2. Understanding the Dataset

Before proceeding with cleaning, it’s important to understand the variables deeply. This helps guide the cleaning and feature engineering process. The subsequent tables detail the types, meaning, and possible values of the variables in the dataset.

**Table 1: Summary of Titanic Dataset Variables**

| Variable      | Type        | Description                           | Values / Range (excluding NaN)           |
|:-------------|:-----------|:-------------------------------------|:----------------------------------------|
| PassengerId  | Numeric    | Unique identifier for each passenger | 1 – 891 (for train dataset)             |
| Survived     | Categorical| Survival status                     | 0 = No, 1 = Yes                          |
| Pclass       | Categorical| Passenger class                      | 1 = 1st, 2 = 2nd, 3 = 3rd               |
| Name         | Text       | Passenger name                       | N/A                                      |
| Sex          | Categorical| Gender                               | Male, Female                              |
| Age          | Numeric    | Age in years                          | 0.42 – 80                                |
| SibSp        | Numeric    | Number of siblings/spouses aboard    | 0 – 8                                    |
| Parch        | Numeric    | Number of parents/children aboard    | 0 – 6                                    |
| Ticket       | Text       | Ticket number                         | N/A                                      |
| Fare         | Numeric    | Passenger fare                        | 0 – 512.33                               |
| Cabin        | Text       | Cabin number                          | N/A (many missing)                        |
| Embarked     | Categorical| Port of embarkation                  | C = Cherbourg, Q = Queenstown, S = Southampton |

**Table 2: Categorical Variables Details**

| Variable   | Unique Values / Categories              | Description                                     |
|:----------|:---------------------------------------|:-----------------------------------------------|
| Survived  | 0, 1                                   | 0 = Did not survive, 1 = Survived              |
| Pclass    | 1, 2, 3                                | Passenger class (1st, 2nd, 3rd)               |
| Sex       | Male, Female                            | Gender                                         |
| Embarked  | C, Q, S                                 | Port of embarkation: Cherbourg, Queenstown, Southampton |


## 4. Deal with missing values

In [10]:
train_df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

Using `.isnull().sum()`, we identified the columns with missing values. They are:

- `Age` with 177 missing values  
- `Cabin` with 687 missing values  
- `Embarked` with 2 missing values  

We address these missing values as follows:

- Imputing numeric missing values in `Age` with the **median** to account for skewness in the distribution  
- Imputing categorical missing values in `Cabin` with `"Unknown"` to indicate missing cabin information  
- Imputing categorical missing values in `Embarked` with the **mode** (`'S'`) to fill the most common port of embarkation  

This approach preserves data consistency while appropriately handling uncertainty in missing values.


In [13]:
train_df['Age'] = train_df['Age'].fillna(train_df['Age'].median())
train_df['Cabin'] = train_df['Cabin'].fillna('Unknown')
train_df['Embarked'] = train_df['Embarked'].fillna(train_df['Embarked'].mode()[0])

In [14]:
train_df.isnull().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

### 5. Removing Duplicates

Duplicates can distort statistical summaries and model performance. Using `.duplicated().sum()`, we count duplicate records



In [15]:
train_df.duplicated().sum()

0

We can confirm that we have no duplicates left in the dataset at this juncture.


In [16]:
train_df.shape

(891, 12)

We also inspect the current shape of the dataset and see that we have *891* rows and *12* columns.

### 6. Standardize Categorical Variables

**Remove leading/trailing spaces and convert strings to lowercase**

To ensure categorical variables are consistent and easy to process, we clean the data by removing any extra spaces and converting all text to lowercase. This step helps avoid inconsistencies such as `"male"`, `" Male"`, or `"MALE"` being treated as different categories.


In [17]:
train_df.columns[(train_df.dtypes == object)]

Index(['Name', 'Sex', 'Ticket', 'Cabin', 'Embarked'], dtype='object')

In [18]:
train_df.columns = train_df.columns.str.lower()
train_df.columns

Index(['passengerid', 'survived', 'pclass', 'name', 'sex', 'age', 'sibsp',
       'parch', 'ticket', 'fare', 'cabin', 'embarked'],
      dtype='object')

In [19]:
train_df

passengerid  survived  pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  name     sex   age  sibsp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female  28.0      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     parch            ticket     fare    cabin embarked  
0        0         A/5 21171   7.2500  Unknown        S  
1        0          PC 17599  71.2833      C85        C  
2        0  STON/O2. 3101282   7.9250  Unknown        S  
3        0            113803  53.1000     C123        S  
4        0            373450   8.0500  Unknown        S  
..     ...               ...      ...      ...      ...  
886      0            211536  13.0000  Unknown        S  
887      0            112053  30.0000      B42        S  
888      2        W./C. 6607  23.4500  Unknown        S  
889      0            111369  30.0000     C148        C  
890      0            370376   7.7500  Unknown        Q  

[891 rows x 12 columns]

Finally, we save the clean, processed dataset as a CSV file in our `processed` directory for future modelling and analysis.


In [20]:
final_file =os.path.join(processed_dir,'titanic_cleaned.csv') 
train_df.to_csv(final_file, index=False)